### DISCUSSION TITLE
bug: strings not accepted in files param

### DISCUSSION THREAD

**Discussion by adriangb** [encode/httpx#1897] [Potential Issue] [answered]
```python
import requests
import httpx

# requests works
resp = requests.post("https://httpbin.org/post", files=[("name", "content")])
assert resp.status_code == 200
assert resp.json()["files"] == {"name": "content"}

# httpx fails, despite the signature of `files` saying it accepts a 2 tuple of (name: str, data: Union[str, bytes, file])
httpx.post("https://httpbin.org/post", files=[("name", "content")])   # raises a TypeError
assert resp.status_code == 200
assert resp.json()["files"] == {"name": "content"}
```

**Comment by lovelydinosaur** [comment, marked as answer]
So... `content` there needs to be bytes - the annotation is derived from here... https://github.com/encode/httpx/blob/master/httpx/_types.py#L84-L93

```python
FileContent = Union[IO[bytes], bytes]
FileTypes = Union[
    # file (or text)
    FileContent,
    # (filename, file (or text))
    Tuple[Optional[str], FileContent],
    # (filename, file (or text), content_type)
    Tuple[Optional[str], FileContent, Optional[str]],
]
RequestFiles = Union[Mapping[str, FileTypes], Sequence[Tuple[str, FileTypes]]]
```

Those comments look like they need updating tho'. They're not in line with what `FileContent` actually accepts.

We switched from accepting text-or-bytes [to bytes-only here](https://github.com/encode/httpx/pull/1736).
We could feasibly reconsider that, so that we only accept files-in-bytes-mode or bytes or text.

(We need to be able to know upfront how many bytes the content will be, and files opened in text mode don't allow us to do that, but text *strings* could be allowable, perhaps.)

**Reply by adriangb** [reply]
Hi Tom, thanks for the quick reply.

That makes sense. I may have been confused myself about the types (as per my comment above, for some reason I thought the type signature did accept `str`). I just opened https://github.com/encode/httpx/pull/1898 to fix the comments.

Regarding the actual functionality, while it would be nice to be compatible w/ requests here, you must have already weighed the pros and cons in #1736 and friends, so I don't think I have anything to add there.

Either way, I'll close this because it's obvious that this is not a bug (beyond the comments being out of date), and a change to the functionality would be a feature request.
